Mathlib/Data/Finset/LocallyFinite/Box.lean

```diff
@@ -0,0 +1,92 @@
+/-
+Copyright (c) 2024 Yaël Dillies. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies
+-/
+import Mathlib.Algebra.Ring.Prod
+import Mathlib.Data.Int.Interval
+import Mathlib.Order.Disjointed
+import Mathlib.Tactic.Ring
+import Mathlib.Tactic.Zify
+
+/-!
+# Decomposing a locally finite ordered ring into boxes
+
+This file proves that any locally finite ordered ring can be decomposed into "boxes", namely
+differences of consecutive intervals.
+
+## Implementation notes
+
+We don't need the full ring structure, only that there is an embedding `ℕ → `
+-/
+
+/-! ### General locally finite ordered ring -/
+
+namespace Finset
+variable {α : Type*} [OrderedRing α] [LocallyFiniteOrder α] [DecidableEq α] {n : ℕ}
+
+/-- Hollow box centered at `0 : α` going from `-n` to `n`. -/
+def box : ℕ → Finset α := disjointed fun n ↦ Icc (-n : α) n
+
+private lemma Icc_neg_mono : Monotone fun n : ℕ ↦ Icc (-n : α) n := by
+  refine fun m n hmn ↦ by apply Icc_subset_Icc <;> simpa using Nat.mono_cast hmn
+
+@[simp] lemma box_zero : (box 0 : Finset α) = {0} := by simp [box]
+
+lemma box_succ_eq_sdiff (n : ℕ) :
+    box (n + 1) = Icc (-n.succ : α) n.succ \ Icc (-n) n := Icc_neg_mono.disjointed_succ _
+
+lemma disjoint_box_succ_prod (n : ℕ) : Disjoint (box (n + 1)) (Icc (-n : α) n) := by
+  rw [box_succ_eq_sdiff]; exact disjoint_sdiff_self_left
+
+@[simp] lemma box_succ_union_prod (n : ℕ) :
+    box (n + 1) ∪ Icc (-n : α) n = Icc (-n.succ : α) n.succ := Icc_neg_mono.disjointed_succ_sup _
+
+@[simp] lemma box_succ_disjUnion (n : ℕ) :
+    (box (n + 1)).disjUnion (Icc (-n : α) n) (disjoint_box_succ_prod _) =
+      Icc (-n.succ : α) n.succ := by rw [disjUnion_eq_union, box_succ_union_prod]
+
+@[simp] lemma zero_mem_box : (0 : α) ∈ box n ↔ n = 0 := by cases n <;> simp [box_succ_eq_sdiff]
+
+end Finset
+
+open Finset
+
+/-! ### Product of locally finite ordered rings -/
+
+namespace Prod
+variable {α β : Type*} [OrderedRing α] [OrderedRing β] [LocallyFiniteOrder α] [LocallyFiniteOrder β]
+  [DecidableEq α] [DecidableEq β] [@DecidableRel (α × β) (· ≤ ·)]
+
+@[simp] lemma card_box_succ (n : ℕ) :
+    (box (n + 1) : Finset (α × β)).card =
+      (Icc (-n.succ : α) n.succ).card * (Icc (-n.succ : β) n.succ).card -
+        (Icc (-n : α) n).card * (Icc (-n : β) n).card := by
+  rw [box_succ_eq_sdiff, card_sdiff (Icc_neg_mono n.le_succ), Prod.card_Icc, Prod.card_Icc]; rfl
+
+end Prod
+
+/-! ### `ℤ × ℤ` -/
+
+namespace Int
+variable {n : ℕ} {x : ℤ × ℤ}
+
+attribute [norm_cast] toNat_ofNat
+
+lemma card_box : ∀ {n}, n ≠ 0 → (box n : Finset (ℤ × ℤ)).card = 8 * n
+  | n + 1, _ => by
+    simp_rw [Prod.card_box_succ, card_Icc, sub_neg_eq_add]
+    norm_cast
+    refine tsub_eq_of_eq_add ?_
+    zify
+    ring
+
```

Comment:
can you add the result that `box 0` has size `1`.

Reply:
No need. simp already knows that: `(box 0).card = {0}.card = 1`